src/awkward/_v2/_connect/cuda/__init__.py

```diff
@@ -0,0 +1,142 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+import os
+import glob
+import math
+import sys
+
+import numpy
+
+import awkward
+
+kernel_specializations = {
+    "awkward_ListArray32_num_64": "cuda_ListArray_num<int32_t, int64_t>",
+    "awkward_ListArrayU32_num_64": "cuda_ListArray_num<uint32_t, int64_t>",
+    "awkward_ListArray64_num_64": "cuda_ListArray_num<int64_t, int64_t>",
+    "awkward_RegularArray_num_64": "cuda_RegularArray_num<int64_t>",
+    "awkward_BitMaskedArray_to_ByteMaskedArray": "cuda_BitMaskedArray_to_ByteMaskedArray",
+    "awkward_ListArray32_validity": "cuda_ListArray_validity<int32_t>",
+    "awkward_ListArrayU32_validity": "cuda_ListArray_validity<uint32_t>",
+    "awkward_ListArray64_validity": "cuda_ListArray_validity<int64_t>",
+}
```

Comment:
Should this be hand-written (and therefore in git) or auto-generated? It looks like something that would be easier to auto-generate.

Also, for the prefix ("`cuda_`" rather than "`awkward_`"), the reason we have a prefix is to avoid collisions in a namespace—suppose that a user is also writing `cupy.RawKernel` and they use the same names. If we don't have actual C++ namespaces with "`::`", we use a prefix. (The cpu-kernels have a C interface, so no namespaces.)

  1. Perhaps you want the prefix to be "`awkward`", rather than "`cuda`", since you're trying to disambiguate these functions from other CUDA functions, but not from other Awkward functions. (One of the benefits of a library with an odd name is that no one else is likely to pick the same thing.)
  2. Can you use real C++ namespaces now that we're JIT-compiling?

For the automated mapping of names, your starting point is what the kernel calls are doing:

```python
nplike["name_of_kernel_without_specialization", np.uint32, np.int64](arguments)
```

In your bookkeeping, you need to get from a kernel name without specialization plus the dtype of every array (pointer-based) argument passed into or out of it. If you arrange your kernels to be templated on all the pointer-based arguments, regardless of whether it needs to be (you might have only one specialization), then the mapping can be automated.

Reply:
There were a lot of ambiguity when I tried to go down this route. I came up with an alternate solution of opening each and every `cpu-kernel` and extracting the specializations from there. There are only a handful of ambiguous cases (5), which can be handwritten when those respective kernels are implemented.

1. Yes, the kernels are now prefixed with `awkward_`.
2. I can use real C++ namespaces now, is `cupy`  a good namespace since these kernels are actually `cupy` RawKernel's?